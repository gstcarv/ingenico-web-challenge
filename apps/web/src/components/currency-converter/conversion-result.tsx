import { useConversionStore } from "../../stores/conversion-store";

export const ConversionResult = () => {
    const { conversionResult } = useConversionStore();

    if (!conversionResult) {
        return null;
    }

    return (
        <div className="mt-6 p-4 bg-success-50 border border-success-200 rounded-lg">
            <h3 className="text-lg font-semibold text-brand-700 mb-2">Conversion Result</h3>
            <div className="space-y-2">
                <div className="flex justify-between items-center">
                    <span className="text-neutral-700 font-medium">Original Amount</span>
                    <span className="text-brand-800 font-bold">
                        {conversionResult.originalAmount} {conversionResult.fromCurrency}
                    </span>
                </div>
                <div className="flex justify-between items-center">
                    <span className="text-neutral-700 font-medium">Converted Amount</span>
                    <span className="text-brand-800 font-bold">
                        {conversionResult.convertedAmount} {conversionResult.toCurrency}
                    </span>
                </div>
                <div className="flex justify-between items-center">
                    <span className="text-neutral-700 font-medium">Exchange Rate</span>
                    <span className="text-brand-800 font-bold">
                        1 {conversionResult.fromCurrency} = {conversionResult.exchangeRate.toFixed(4)}{" "}
                        {conversionResult.toCurrency}
                    </span>
                </div>
                <div className="text-sm text-neutral-600 mt-2">
                    Date: {conversionResult.date} | Last updated:{" "}
                    {new Date(conversionResult.lastUpdatedAt).toLocaleString()}
                </div>
            </div>
        </div>
    );
};
