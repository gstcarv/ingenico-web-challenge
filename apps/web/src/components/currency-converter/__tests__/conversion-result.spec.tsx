import { render, screen } from "@testing-library/react";
import { describe, expect, it, vi, beforeEach, afterEach } from "vitest";
import { ConversionResult } from "../conversion-result";
import type { ProcessedConversionResponse } from "../../../types/currency";

// Mock the conversion store
vi.mock("../../../stores/conversion-store", () => ({
    useConversionStore: vi.fn(),
}));

const mockUseConversionStore = vi.mocked(await import("../../../stores/conversion-store")).useConversionStore;

describe("ConversionResult", () => {
    const mockConversionResult: ProcessedConversionResponse = {
        originalAmount: 100,
        convertedAmount: 85.5,
        fromCurrency: "USD",
        toCurrency: "EUR",
        exchangeRate: 0.855,
        date: "2024-01-15",
        lastUpdatedAt: "2024-01-15T10:30:00Z",
    };

    beforeEach(() => {
        vi.clearAllMocks();
        // Mock Date.prototype.toLocaleString to return a consistent date string
        vi.spyOn(Date.prototype, "toLocaleString").mockReturnValue("1/15/2024, 7:30:00 AM");
    });

    afterEach(() => {
        vi.restoreAllMocks();
    });

    it("should render nothing when conversionResult is null", () => {
        mockUseConversionStore.mockReturnValue({
            conversionResult: null,
        });

        const { container } = render(<ConversionResult />);
        expect(container.firstChild).toBeNull();
    });

    it("should render conversion result with correct structure", () => {
        mockUseConversionStore.mockReturnValue({
            conversionResult: mockConversionResult,
        });

        render(<ConversionResult />);

        expect(screen.getByText("Conversion Result")).toBeInTheDocument();
        expect(screen.getByText("Original Amount")).toBeInTheDocument();
        expect(screen.getByText("Converted Amount")).toBeInTheDocument();
        expect(screen.getByText("Exchange Rate")).toBeInTheDocument();
    });

    it("should format currency values correctly", () => {
        mockUseConversionStore.mockReturnValue({
            conversionResult: mockConversionResult,
        });

        render(<ConversionResult />);

        // Check original amount formatting
        expect(screen.getByText("$100.00")).toBeInTheDocument();

        // Check converted amount formatting
        expect(screen.getByText("€85.50")).toBeInTheDocument();
    });

    it("should format exchange rate correctly", () => {
        mockUseConversionStore.mockReturnValue({
            conversionResult: mockConversionResult,
        });

        render(<ConversionResult />);

        expect(screen.getByText("1 USD = 0.8550 EUR")).toBeInTheDocument();
    });

    it("should display date and last updated information", () => {
        mockUseConversionStore.mockReturnValue({
            conversionResult: mockConversionResult,
        });

        render(<ConversionResult />);

        expect(screen.getByText(/Date: 2024-01-15/)).toBeInTheDocument();
        expect(screen.getByText(/Last updated:/)).toBeInTheDocument();
    });

    it("should handle different currency codes correctly", () => {
        const cryptoConversionResult: ProcessedConversionResponse = {
            originalAmount: 1,
            convertedAmount: 45000,
            fromCurrency: "BTC",
            toCurrency: "USD",
            exchangeRate: 45000,
            date: "2024-01-15",
            lastUpdatedAt: "2024-01-15T10:30:00Z",
        };

        mockUseConversionStore.mockReturnValue({
            conversionResult: cryptoConversionResult,
        });

        render(<ConversionResult />);

        expect(screen.getByText("BTC 1.00")).toBeInTheDocument();
        expect(screen.getByText("$45,000.00")).toBeInTheDocument();
        expect(screen.getByText("1 BTC = 45,000.0000 USD")).toBeInTheDocument();
    });

    it("should handle decimal precision correctly", () => {
        const preciseConversionResult: ProcessedConversionResponse = {
            originalAmount: 1.123456,
            convertedAmount: 0.987654,
            fromCurrency: "USD",
            toCurrency: "JPY",
            exchangeRate: 0.987654,
            date: "2024-01-15",
            lastUpdatedAt: "2024-01-15T10:30:00Z",
        };

        mockUseConversionStore.mockReturnValue({
            conversionResult: preciseConversionResult,
        });

        render(<ConversionResult />);

        expect(screen.getByText("$1.123456")).toBeInTheDocument();
        expect(screen.getByText("¥0.987654")).toBeInTheDocument();
        expect(screen.getByText("1 USD = 0.987654 JPY")).toBeInTheDocument();
    });

    it("should have correct CSS classes and styling", () => {
        mockUseConversionStore.mockReturnValue({
            conversionResult: mockConversionResult,
        });

        const { container } = render(<ConversionResult />);

        const mainContainer = container.firstChild as HTMLElement;
        expect(mainContainer).toHaveClass("mt-6", "p-4", "bg-success-50", "border", "border-success-200", "rounded-lg");
    });

    it("should handle zero amounts correctly", () => {
        const zeroConversionResult: ProcessedConversionResponse = {
            originalAmount: 0,
            convertedAmount: 0,
            fromCurrency: "USD",
            toCurrency: "EUR",
            exchangeRate: 0,
            date: "2024-01-15",
            lastUpdatedAt: "2024-01-15T10:30:00Z",
        };

        mockUseConversionStore.mockReturnValue({
            conversionResult: zeroConversionResult,
        });

        render(<ConversionResult />);

        expect(screen.getByText("$0.00")).toBeInTheDocument();
        expect(screen.getByText("€0.00")).toBeInTheDocument();
        expect(screen.getByText("1 USD = 0.0000 EUR")).toBeInTheDocument();
    });
});
