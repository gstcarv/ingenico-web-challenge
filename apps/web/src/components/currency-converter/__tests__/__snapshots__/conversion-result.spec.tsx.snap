// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ConversionResult > should match snapshot 1`] = `
<div
  class="mt-6 p-4 bg-success-50 border border-success-200 rounded-lg"
>
  <h3
    class="text-lg font-semibold text-brand-700 mb-2"
  >
    Conversion Result
  </h3>
  <div
    class="space-y-2"
  >
    <div
      class="flex justify-between items-center"
    >
      <span
        class="text-neutral-700 font-medium"
      >
        Original Amount
      </span>
      <span
        class="text-brand-800 font-bold"
      >
        $100.00
      </span>
    </div>
    <div
      class="flex justify-between items-center"
    >
      <span
        class="text-neutral-700 font-medium"
      >
        Converted Amount
      </span>
      <span
        class="text-brand-800 font-bold"
      >
        â‚¬85.50
      </span>
    </div>
    <div
      class="flex justify-between items-center"
    >
      <span
        class="text-neutral-700 font-medium"
      >
        Exchange Rate
      </span>
      <span
        class="text-brand-800 font-bold"
      >
        1 USD = 0.8550 EUR
      </span>
    </div>
    <div
      class="text-sm text-neutral-600 mt-2"
    >
      Date: 
      2024-01-15
       | Last updated:
       
      15/01/2024, 07:30:00
    </div>
  </div>
</div>
`;
